lib/src/main/java/io/ably/lib/http/AsyncHttpScheduler.java

```diff
@@ -6,27 +6,49 @@
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 
+import io.ably.lib.util.Log;
+
 /**
  * A HttpScheduler that uses a thread pool to run HTTP operations.
  */
-public class AsyncHttpScheduler extends HttpScheduler<ThreadPoolExecutor> {
+public class AsyncHttpScheduler extends HttpScheduler {
     public AsyncHttpScheduler(HttpCore httpCore, ClientOptions options) {
-        super(httpCore, new ThreadPoolExecutor(options.asyncHttpThreadpoolSize, options.asyncHttpThreadpoolSize, KEEP_ALIVE_TIME, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>()));
-        executor.allowsCoreThreadTimeOut();
-    }
-
-    public void dispose() {
-        ThreadPoolExecutor threadPoolExecutor = executor;
-        threadPoolExecutor.shutdown();
-        try {
-            threadPoolExecutor.awaitTermination(SHUTDOWN_TIME, TimeUnit.MILLISECONDS);
-        } catch (InterruptedException e) {
-            threadPoolExecutor.shutdownNow();
-        }
+        super(httpCore, new CloseableThreadPoolExecutor(options));
     }
 
     private static final long KEEP_ALIVE_TIME = 2000L;
-    private static final long SHUTDOWN_TIME = 5000L;
 
     protected static final String TAG = AsyncHttpScheduler.class.getName();
+
+    private static class CloseableThreadPoolExecutor implements CloseableExecutor {
+        private final ThreadPoolExecutor executor;
+
+        CloseableThreadPoolExecutor(final ClientOptions options) {
+            executor = new ThreadPoolExecutor(
+                options.asyncHttpThreadpoolSize,
+                options.asyncHttpThreadpoolSize,
+                KEEP_ALIVE_TIME,
+                TimeUnit.MILLISECONDS,
+                new LinkedBlockingQueue<Runnable>()
+            );
+        }
+
+        @Override
+        public void execute(final Runnable command) {
+            executor.execute(command);
+        }
+
+        @Override
+        public void close() throws Exception {
+            final int drainedCount = executor.shutdownNow().size();
```

Comment:
Will this not shut down all running threads in the thread pool? Isn't that we only want to shut down waiting threads 🤔

Reply:
Yes, that's the intention. I'm trying to keep the implementation as simple as possible. There was [some implementation previously (unused, as it happened)](https://github.com/ably/ably-java/pull/807/commits/6705060846c3a9f8b2102528476afa5bac157fb4) which appeared to want to try to 'gracefully' shut things down - but that is a risky strategy, especially when this API is synchronous and there's no clear structure as to who/what would hold a strong reference to things while they close down anyway. Therefore, I've taken the view that an explicit `close()` like this means that the user is not expecting any further responses back from this client instance.

Reply:
Sorry if I am missing the point of usage of this class. So for example when there are some threads that are running (eg. sending requests) will there be a case that they would be shutdown before they got a chance to complete?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#shutdown()
From shutdownNow doc
> Attempts to stop all actively executing tasks, halts the processing of waiting tasks, and returns a list of the tasks that were awaiting execution.

My question is probably be better formed as - What would happen to those tasks that were awaiting execution?

Reply:
> So for example when there are some threads that are running (eg. sending requests) will there be a case that they would be shutdown before they got a chance to complete?

Correct. That is my intention.